src/main/java/org/jetbrains/bsp/bazel/server/BepServer.java

```diff
@@ -193,68 +189,69 @@ private void processCompletedEvent(BuildEventStreamProtos.TargetComplete targetC
-                throw new RuntimeException("Wrong parts in sketchy textproto parsing: " + parts);
+            // TODO(gerardd) the body of try-catch will be extracted
+            try {
+              List<String> lines =
+                  com.google.common.io.Files.readLines(
+                      new File(protoPathUri), StandardCharsets.UTF_8);
+
+              for (String line : lines) {
+                List<String> parts = Splitter.on("\"").splitToList(line);
+                if (parts.size() != 3) {
+                  throw new RuntimeException("Wrong parts in sketchy textproto parsing: " + parts);
+                }
+
+                compilerClasspath.add(
+                    Uri.fromExecPath(
+                        "exec-root://" + parts.get(1), bspServer.getBazelData().getExecRoot()));
               }
-
-              compilerClasspath.add(
-                  Uri.fromExecPath(
-                      "exec-root://" + parts.get(1), bspServer.getBazelData().getExecRoot()));
+            } catch (IOException e) {
+              throw new RuntimeException(e);
             }
           }
         }
       }
     }
   }
 
-  private void processActionEventIfPresent(BuildEventStreamProtos.BuildEvent event)
-      throws IOException {
+  private void processActionEventIfPresent(BuildEventStreamProtos.BuildEvent event) {
     if (event.hasAction()) {
       processActionEvent(event.getAction());
     }
   }
 
-  private void processActionEvent(BuildEventStreamProtos.ActionExecuted action) throws IOException {
+  private void processActionEvent(BuildEventStreamProtos.ActionExecuted action) {
     if (!Constants.SUPPORTED_COMPILERS.contains(action.getType())) {
       // Ignore file template writes and such.
       // TODO(illicitonion): Maybe include them as task notifications (rather than diagnostics).
       return;
     }
 
     BuildTargetIdentifier target = new BuildTargetIdentifier(action.getLabel());
-    boolean hasDiagnosticsOutput = diagnosticsProtosLocations.containsKey(target.getUri());
-    Map<Uri, List<PublishDiagnosticsParams>> filesToDiagnostics = new HashMap<>();
-
-    for (BuildEventStreamProtos.File log : action.getActionMetadataLogsList()) {
-      if (!log.getName().equals("diagnostics")) {
-        continue;
-      }
 
-      LOGGER.info("Found diagnostics file in {}", log.getUri());
-
-      if (hasDiagnosticsOutput) {
-        diagnosticsProtosLocations.remove(target.getUri());
+    Map<Uri, List<PublishDiagnosticsParams>> filesToDiagnostics =
+        action.getActionMetadataLogsList().stream()
+            .filter(log -> log.getName().equals("diagnostics"))
+            .peek(log -> LOGGER.info("Found diagnostics file in {}", log.getUri()))
+            .map(log -> diagnosticsDispatcher.collectDiagnostics(target, log.getUri().substring(7)))
+            .collect(HashMap::new, Map::putAll, Map::putAll);
+
+    if (hasDiagnosticsOutput(target)) {
+      if (filesToDiagnostics.isEmpty()) {
```

Comment:
Why did you separate that if with && into two ifs? This changed the semantics https://github.com/JetBrains/bazel-bsp/blob/dev/src/main/java/org/jetbrains/bsp/bazel/server/BepServer.java#L292

Reply:
It's not the only change there. I talked [about it](https://github.com/JetBrains/bazel-bsp/pull/21#discussion_r534587407) with André.